commit.go

```diff
@@ -0,0 +1,432 @@
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package main
+
+import (
+	"bytes"
+	"context"
+	"crypto/rand"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"runtime"
+	"strings"
+	"time"
+
+	"github.com/AkihiroSuda/nerdctl/pkg/idutil"
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/cio"
+	"github.com/containerd/containerd/content"
+	"github.com/containerd/containerd/diff"
+	"github.com/containerd/containerd/errdefs"
+	"github.com/containerd/containerd/images"
+	"github.com/containerd/containerd/leases"
+	"github.com/containerd/containerd/platforms"
+	refdocker "github.com/containerd/containerd/reference/docker"
+	"github.com/containerd/containerd/rootfs"
+	"github.com/containerd/containerd/snapshots"
+	"github.com/opencontainers/go-digest"
+	"github.com/opencontainers/image-spec/identity"
+	"github.com/opencontainers/image-spec/specs-go"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"github.com/urfave/cli/v2"
+)
+
+type commitOpts struct {
+	author  string
+	message string
+	ref     string
+}
+
+var (
+	commitCommand = &cli.Command{
+		Name:        "commit",
+		Usage:       "[flags] CONTAINER REPOSITORY[:TAG]",
+		Description: "Create a new image from a container's changes",
+		Action:      commitAction,
+		Flags: []cli.Flag{
+			&cli.StringFlag{
+				Name:    "author",
+				Aliases: []string{"a"},
+				Usage:   `Author (e.g., "nerdctl contributor <nerdctl-dev@a-team.com>")`,
+			},
+			&cli.StringFlag{
+				Name:    "message",
+				Aliases: []string{"m"},
+				Usage:   `Commit message`,
+			},
+		},
+	}
+
+	emptyGZLayer = digest.Digest("sha256:4f4fb700ef54461cfa02571ae0db9a0dc1e0cdb5577484a6d75e68dc38e8acc1")
+)
+
+func commitAction(clicontext *cli.Context) error {
+	if clicontext.NArg() != 2 {
+		return errors.New("need container and commit image name")
+	}
+
+	opts, err := newCommitOpts(clicontext)
+	if err != nil {
+		return err
+	}
+
+	client, ctx, cancel, err := newClient(clicontext)
+	if err != nil {
+		return err
+	}
+	defer cancel()
+
+	return idutil.WalkContainers(ctx, client, []string{clicontext.Args().First()}, func(ctx context.Context, client *containerd.Client, _, id string) error {
+		imageID, err := commitContainer(ctx, client, id, opts)
+		if err != nil {
+			return err
+		}
+
+		_, err = fmt.Fprintf(clicontext.App.Writer, "%s\n", imageID)
+		return err
+	})
+}
+
+func newCommitOpts(clicontext *cli.Context) (*commitOpts, error) {
+	rawRef := clicontext.Args().Get(1)
+
+	named, err := refdocker.ParseDockerRef(rawRef)
+	if err != nil {
+		return nil, err
+	}
+
+	return &commitOpts{
+		author:  clicontext.String("author"),
+		message: clicontext.String("message"),
+		ref:     named.String(),
+	}, nil
+}
+
+var emptyDigest = digest.Digest("")
+
+func commitContainer(ctx context.Context, client *containerd.Client, id string, opts *commitOpts) (digest.Digest, error) {
```

Comment:
Can we split this as a separate package? Like pkg/imgutil/commit

Reply:
Sure. But I need to do some refactor for this. Could I handle it in the followup?

Reply:
opened https://github.com/AkihiroSuda/nerdctl/pull/43